admin-src/DiagnoseResponse/index.js

```diff
@@ -8,52 +8,107 @@ import { firebaseTimestampToMoment } from '../utils/date'
+
+        if (changeType === 'added') {
+          if (!newDiagnoses.find(diagnose => diagnose.id === doc.id)) {
+            const docData = doc.data()
+            const userUID = getUserUIDFromDiagnoseRef(doc.ref)
+            const userSnap = await firebase.firestore().collection('users').doc(userUID).get()
+            const username = userSnap.get('username')
+
+            console.log(docData)
+            newDiagnoses.push({
+              id: doc.id,
+              ref: doc.ref,
+              userUID,
+              username,
+              ...docData
             })
-          )
+          }
+        }
 
-          setDiagnoses([...unanswered])
-        })
+        if (changeType === 'removed') {
+          newDiagnoses = newDiagnoses.filter(diagnose => diagnose.id !== doc.id)
+        }
+      })
+    )
+
+    if (filter) {
+      newDiagnoses = newDiagnoses.filter(filter)
+    }
+
+    setDiagnoses([...newDiagnoses])
+  }
+
+  const unansweredQuery = () => {
+    return firebase
+      .firestore()
+      .collectionGroup('requests')
+      .where('amountOfAnswers', '==', 0)
+      .onSnapshot(onSnapshot)
+  }
+  const dateDaysAgo = (daysAgo) => {
+    const date = new Date(Date.now())
+    date.setDate(date.getDate() - daysAgo)
+    return date
+  }
+  const filterBySolved = (diagnose) => diagnose.solved
+  const filterByNotSolved = (diagnose) => !diagnose.solved
+  const filterByLastActivity = (diagnose) => (diagnose.updatedAt !== undefined) && diagnose.updatedAt.toDate() >= dateDaysAgo(STALE_STATUS_AFTER_DAYS)
+  const filterByAmountOfAnswers = (diagnose) => diagnose.amountOfAnswers > 0
+  const filterStale = (diagnose) => !filterByLastActivity(diagnose) && filterByNotSolved(diagnose) && filterByAmountOfAnswers(diagnose)
+  const filterInDiscussion = (diagnose) => filterByLastActivity(diagnose) && filterByNotSolved(diagnose) && filterByAmountOfAnswers(diagnose)
+  const inDiscussionQuery = () => {
+    return firebase
+      .firestore()
+      .collectionGroup('requests')
+      .where('amountOfAnswers', '>', 0)
+      .onSnapshot((snapshot) => onSnapshot(snapshot, filterInDiscussion))
+  }
+
+  const staleQuery = () => {
+    return firebase
+      .firestore()
+      .collectionGroup('requests')
+      .where('amountOfAnswers', '>', 0)
+      .onSnapshot((snapshot) => onSnapshot(snapshot, filterStale))
+  }
+
+  const solvedQuery = () => {
+    return firebase
+      .firestore()
+      .collectionGroup('requests')
+      .onSnapshot((snapshot) => onSnapshot(snapshot, filterBySolved))
+  }
+
+  useEffect(() => {
+    const attachQueryListener = () => {
+      if (filter === 'unanswered') {
+        unansweredQuery()
+      } else if (filter === 'in discussion') {
+        inDiscussionQuery()
+      } else if (filter === 'stale') {
+        staleQuery()
+      } else if (filter === 'solved') {
+        solvedQuery()
+      }
     }
 
-    attachQueryListenerForUnansweredDiagnoses()
+    attachQueryListener()
   }, [])
```

Comment:
I think it would be better if `filter` is actually the function that filters the data, not a string. It's not a good design choice to use strings to select options. It's better to make the component open so that any filter can be applied.

Reply:
I think what is wrong is the name of the variable it should not be 'filter' but 'state' instead that's why it lead. Don't forget almost all react-native components use strings in select options.

Reply:
That React Native uses this pattern doesn't mean it's correct. You're breaking the open/closed principle here. A component should be open to extension but closed to modification. Consider what happens if we need to change/remove/add a filter:
- **With your solution:** we would need to _modify the component_ to change/remove/add the filter.
- **With my suggestion:** you only have to create the new filter and pass that as a prop. This is called dependency inversion/injection if I remember correctly.

cc @MPucheta what do you think?

Reply:
I agree with Nahuel's suggestion. These "strings" options are almost always solved by using some form of polymorphism too (I do not like explicit inheritance hierarchy, but you could have all these filter follow the same interface and achieve the same effect, for example, all of them receive a snapshot and a filter) I would also try to separate all those filters and queries into separate files.

Reply:
I agree with your comments, what you are suggesting is obviously a better solution object oriented. It also would be better to not pass any filter but the diagnose objects instead. I just applied this solution for the given delivery date.  I'll make the changes